### PR TITLE
overlay: add layer visibility helpers (CRAFT-477)

### PR DESCRIPTION
**Pull request by cmatsuoka**
Add helper functions to obtain the sets of migratable files and
directories from each overlay layer considering whited out entries and
opaque directories.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

File: tests/unit/overlays/test_overlays.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+class TestHelpers:
+    """OCI special files translation and verification."""
+
+    def test_is_oci_opaque_dir(self, new_dir):
+        dir1 = Path("dir1")
+        dir1.mkdir()
+        Path("dir1/.wh..wh..opq").touch()
+
+        assert overlays.is_oci_opaque_dir(dir1)
```
Comment (cjp256) [inline tests/unit/overlays/test_overlays.py:32]:
```suggestion
        assert overlays.is_oci_opaque_dir(dir1) is True
```

File: tests/unit/overlays/test_overlays.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+    def test_is_oci_opaque_dir_symlink(self, new_dir):
+        dir1 = Path("dir1")
+        dir1.mkdir()
+        Path("dir1/.wh..wh..opq").touch()
+
+        dir2 = Path("dir2")
+        dir2.symlink_to(dir1)
+
+        assert overlays.is_oci_opaque_dir(dir1)
```
Comment (cjp256) [inline tests/unit/overlays/test_overlays.py:48]:
```suggestion
        assert overlays.is_oci_opaque_dir(dir1) is True
```

File: tests/unit/overlays/test_overlays.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        [("foo", "foo/.wh..wh..opq"), ("/path/foo", "/path/foo/.wh..wh..opq")],
+    )
+    def test_oci_opaque_dir(self, name, oci_name):
+        assert overlays.oci_opaque_dir(Path(name)) == Path(oci_name)
+
+    def test_is_path_visible(self, new_dir):
+        deepfile = Path("dir1/dir2/dir3/dir4/foo")
+        deepfile.parent.mkdir(parents=True)
+        deepfile.touch()
+        assert overlays._is_path_visible(new_dir, deepfile)
```
Comment (cjp256) [inline tests/unit/overlays/test_overlays.py:68]:
```suggestion
        assert overlays._is_path_visible(new_dir, deepfile) is True
```

File: craft_parts/overlays/overlays.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    :returns: Whether the given path is an overlayfs opaque directory.
+    """
+    if not path.is_dir() or path.is_symlink():
+        return False
+
+    return oci_opaque_dir(path).exists()
+
+
+def oci_whiteout(path: Path) -> Path:
```
Comment (cjp256) [inline craft_parts/overlays/overlays.py:108]:
```suggestion
def get_oci_whiteout_for_file(path: Path) -> Path:
```

File: craft_parts/overlays/overlays.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    """Convert the given path to an OCI whiteout file name.
+
+    :param path: The file path to white out.
+
+    :returns: The corresponding OCI whiteout file name.
+    """
+    return path.parent / (".wh." + path.name)
+
+
+def oci_opaque_dir(path: Path) -> Path:
```
Comment (cjp256) [inline craft_parts/overlays/overlays.py:118]:
```suggestion
def get_oci_opaque_whiteout_for_directory(path: Path) -> Path:
```

File: craft_parts/overlays/overlays.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Overlay handling helpers."""
```
Comment (cjp256) [inline craft_parts/overlays/overlays.py:17]:
```suggestion
"""Overlay handling helpers.

Relevant OCI documentation available at:
https://github.com/opencontainers/image-spec/blob/main/layer.md
"""
```

File: craft_parts/overlays/overlays.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+import logging
+import os
+from pathlib import Path
+from typing import Set, Tuple
+
+logger = logging.getLogger(__name__)
+
+
+def visible_in_layer(srcdir: Path, destdir: Path) -> Tuple[Set[str], Set[str]]:
```
Comment (cjp256) [inline craft_parts/overlays/overlays.py:27]: src/dst seems a bit unusual in this context - maybe rewrite as lower/upper?
```suggestion
def visible_in_layer(lower_dir: Path, upper_dir: Path) -> Tuple[Set[str], Set[str]]:
```

File: tests/unit/overlays/test_overlays.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        assert files == {"a", "dir1/b", "dir1/c"}
+        assert dirs == {"dir1"}
+
+    def test_visible_in_layer_whited_out_file(self, new_dir):
+        Path("destdir/dir1").mkdir()
+        Path("destdir/dir1/.wh.b").touch()
+
+        files, dirs = overlays.visible_in_layer(Path("srcdir"), Path("destdir"))
+        assert files == {"a", "dir1/c"}
+        assert dirs == set()
```
Comment (cjp256) [inline tests/unit/overlays/test_overlays.py:117]: I imagine this is a bug:
```suggestion
        assert dirs == {"dir1"}
```

File: tests/unit/overlays/test_overlays.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+    def test_visible_in_layer_opaque_dir(self, new_dir):
+        Path("destdir/dir1").mkdir()
+        Path("destdir/dir1/.wh..wh..opq").touch()
+
+        files, dirs = overlays.visible_in_layer(Path("srcdir"), Path("destdir"))
+        assert files == {"a"}
+        assert dirs == set()
```
Comment (cjp256) [inline tests/unit/overlays/test_overlays.py:125]: This too from my understanding of the spec:
```suggestion
        assert dirs == {"dir1"}
```

File: tests/unit/overlays/test_overlays.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+    def test_visible_in_layer_whiteout_in_opaque_dir(self, new_dir):
+        Path("destdir/dir1").mkdir()
+        Path("destdir/dir1/.wh..wh..opq").touch()
+        Path("destdir/dir1/.wh.b").touch()
+
+        files, dirs = overlays.visible_in_layer(Path("srcdir"), Path("destdir"))
+        assert files == {"a"}
+        assert dirs == set()
```
Comment (cjp256) [inline tests/unit/overlays/test_overlays.py:134]: same
```suggestion
        assert dirs == {"dir1"}
```

File: craft_parts/overlays/overlays.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+                directories.remove(directory)
+
+    logger.debug("files=%r, dirs=%r", visible_files, visible_dirs)
+    return visible_files, visible_dirs
+
+
+def _is_path_visible(root: Path, relpath: Path) -> bool:
+    """Verify if any element of the given path is not whited out.
+
+    :var root: The root directory, not included in the verification.
```
[the remark this answers is not in the document]
Comment (cmatsuoka) [reply craft_parts/overlays/overlays.py:78]: Oops. I think `var` stayed in my brain after I added the `ProjectDirs` instance variables documentation :)

File: tests/unit/overlays/test_overlays.py (hunk 1 above)
Comment (cmatsuoka) [reply tests/unit/overlays/test_overlays.py:32]: Done.

**Review by cjp256 — APPROVED** [review]
LGTM THANKS!

Might need to pin pyright
```
+ pyright@1.1.164
added 1 package from 1 contributor in 1.034s
pyright setup.py craft_parts tests
Loading configuration file at /home/runner/work/craft-parts/craft-parts/pyrightconfig.json
Assuming Python version 3.8
Assuming Python platform Linux
stubPath /home/runner/work/craft-parts/craft-parts/typings is not a valid directory.
Searching for source files
Found 159 source files
/home/runner/work/craft-parts/craft-parts/tests/unit/plugins/test_base.py
  /home/runner/work/craft-parts/craft-parts/tests/unit/plugins/test_base.py:42:24 - error: Expression of type "FooPluginProperties" cannot be assigned to declared type "Type[PluginProperties]"
    Type "FooPluginProperties" cannot be assigned to type "Type[PluginProperties]" (reportGeneralTypeIssues)
```

**Pull request merged**